### PR TITLE
♻️ graceful background task stop

### PR DESCRIPTION
**Pull request by GitHK**
## What do these changes do?

An optimisation laying around a PR. That I did not want to toss away.

- ♻️ Always try to stop tasks gracefully before cancelling them if a timeout is reached.
- ♻️ bonus: updated signage of `lock_context`.

## Dev Checklist

- [x] No ENV changes or I properly updated ENV ([read the instruction](https://git.speag.com/oSparc/osparc-ops-deployment-configuration/-/blob/configs/README.md?ref_type=heads#how-to-update-env-variables))

### CODE REVIEW

File: packages/service-library/src/servicelib/background_task.py [hunk 1]
```diff
@@ -23,19 +23,41 @@ class PeriodicTaskCancellationError(PydanticErrorMixin, Exception):
     msg_template: str = "Could not cancel task '{task_name}'"
 
 
+class ContinueCondition:
```
Comment (sanderegg) [inline packages/service-library/src/servicelib/background_task.py:26]: why not use a dataclass decorator here?

File: packages/service-library/src/servicelib/background_task.py [hunk 2]
```diff
@@ -23,19 +23,41 @@ class PeriodicTaskCancellationError(PydanticErrorMixin, Exception):
...
+
+    @property
+    def can_continue(self) -> bool:
+        return self._can_continue
+
+
+class _ExtendedTask(asyncio.Task):
+    def __init__(self, coro, *, loop=None, name=None):
+        super().__init__(coro=coro, loop=loop, name=name)
+        self.continue_condition: ContinueCondition | None = None
```
Comment (sanderegg) [inline packages/service-library/src/servicelib/background_task.py:41]: why is this optional? I don't see it that you optionally put it in???

File: packages/service-library/src/servicelib/background_task.py [hunk 3]
```diff
@@ -99,9 +128,21 @@ async def stop_periodic_task(
...
         logger,
         logging.DEBUG,
-        msg=f"cancel periodic background task '{asyncio_task.get_name()}'",
+        msg=f"stop periodic background task '{asyncio_task.get_name()}'",
     ):
-        await cancel_task(asyncio_task, timeout=timeout)
+        asyncio_task = cast(_ExtendedTask, asyncio_task)
+        continue_condition: ContinueCondition | None = asyncio_task.continue_condition
+        if continue_condition:
+            continue_condition.stop()
```
Comment (sanderegg) [inline packages/service-library/src/servicelib/background_task.py:136]: ok what is the rationale here?
I see some major differences here:
- cancel directly cancels a task vs stop needs to wait for an iteration to complete (an iteration where we have no clue how long it takes), 
- therefore now you wait 2x the timeout in the worst case, how are the services coping with this did you check? I think we do not have the test anymore that would check starting/stopping services to see if they return a 0 exit code

**Review by sanderegg — COMMENTED** [review]
I am not sure I get the cause and the possible consequences of this change.
are you sure it must go in right now?

File: packages/service-library/tests/test_background_task.py [hunk 4]
```diff
@@ -127,4 +127,5 @@ async def test_periodic_task_context_manager(
         await asyncio.sleep(5 * task_interval.total_seconds())
         assert asyncio_task.cancelled() is False
         assert asyncio_task.done() is False
-    assert asyncio_task.cancelled() is True
+    # NOTE: task is no longer cancelled but gracefully stopped
+    assert asyncio_task.cancelled() is False
```
Comment (pcrespov) [inline packages/service-library/tests/test_background_task.py:131]: This is a bit confusing. So your task is neither `done` nor `canceled`? Is it pending? but wait, you are outside of the context manager? What is the state of the task?

**Review by matusdrobuliak66 — APPROVED** [review]

**Comment by GitHK** [comment]
abandoned

**Pull request closed without merging**